### PR TITLE
Implement Save/Approve workflow for GRN receiving - Closes #14682

### LINKED ISSUE #14682: Save Button Required in GRN Receive View
Save Button Required in GRN Receive View

### PR DESCRIPTION
**Pull request by buddhika75**
## Summary
- Added Save and Approve buttons to GRN receiving workflow similar to Purchase Order creation
- Implemented separate save and approve functionality for both regular and costing-enabled environments
- Users can now save GRNs in draft state and finalize them later
- Maintained all existing immediate settle workflows without disruption

### Test Environment Setup
- Ensure both `Manage Costing = true` and `Manage Costing = false` configurations are tested
- Test with different institutions and departments
- Verify purchase orders are available for receiving

#### 1. New GRN Creation (Non-Costing)
**Test Steps:**
1. Navigate to Pharmacy → Purchase Orders → Goods Received Note
2. Click "Save/Approve" button for any pending PO (when Manage Costing = false)
3. Enter receiving details (quantities, batch numbers, expiry dates)
4. Click "Save" button
5. Verify GRN appears in the list with "View/Edit" buttons
6. Click "Edit" to modify the saved GRN
7. Make changes and click "Approve" to finalize

**Expected Results:**
- Save creates draft GRN without validation errors
- Saved GRNs appear in PO list with Edit option
- Edit loads all previously entered data
- Approve finalizes GRN with stock updates and payments

#### 2. New GRN Creation (With Costing)
**Test Steps:**
1. Set Manage Costing = true in configuration
2. Navigate to Pharmacy → Purchase Orders → Goods Received Note  
3. Click "Save/Approve" button for any pending PO
4. Enter receiving details with costing information
5. Verify calculations update in real-time
6. Test Save and Approve workflow

**Expected Results:**
- Costing calculations work correctly
- Purchase rates, retail rates, and totals calculate properly
- Save/Approve functions work with costing enabled

#### 3. Data Persistence Tests
**Critical Test:**
1. Create new GRN and enter complete item details
2. Click Save
3. Navigate away from the page
4. Return to edit the saved GRN
5. Verify ALL data is preserved:
   - Item quantities and free quantities
   - Batch numbers and expiry dates
   - Purchase rates and retail rates
   - Invoice number and date
   - All calculated totals

#### 4. Validation Tests
**Save (Draft) Mode:**
- Should allow saving with incomplete data
- Should not require batch details or prices
- Should not validate mandatory fields

**Approve (Final) Mode:**
- Should validate all required fields
- Should require invoice number
- Should check batch details and prices
- Should update stock and create payments

#### 5. Stock and Financial Impact
**After Approve:**
1. Verify stock quantities are updated correctly
2. Check stock history entries are created
3. Confirm payment records are generated  
4. Validate financial balance updates
5. Ensure Purchase Order status reflects completion

### UI/UX Verification
- Save/Approve buttons appear only for appropriate POs
- View/Edit buttons show based on GRN status
- Button rendering respects Manage Costing configuration
- Real-time calculations work during data entry
- All totals display correctly in both modes

### Regression Testing
- Verify existing immediate settle workflow still works
- Confirm existing GRN pages function normally
- Test that Purchase Order workflows are unaffected
- Validate other pharmacy modules are not impacted

### Edge Cases
- Test with zero quantities
- Test with only free quantities  
- Test with partial receiving scenarios
- Test with different payment methods
- Test concurrent user scenarios

### Performance Testing
- Verify page load times for saved GRNs
- Test calculation performance with multiple items
- Check database query efficiency for large datasets

## Configuration Notes
The feature automatically detects the Manage Costing setting and shows appropriate UI:
- `Manage Costing = false`: Uses regular GRN save/approve workflow
- `Manage Costing = true`: Uses costing-enabled GRN save/approve workflow

## Database Impact
- New GRN records with PHARMACY_GRN_PRE status for saved drafts
- Standard PHARMACY_GRN status for approved/finalized GRNs
- Proper bill item relationships maintained
- Stock and financial transactions only on approval

🤖 Generated with [Claude Code](https://claude.ai/code)

Co-Authored-By: Claude <noreply@anthropic.com>

## Summary by CodeRabbit

- **New Features**
  - Save & Approve workflows for GRNs (costing and non-costing), print-preview, edit-saved GRN flows, and a new GRN costing page.

- **Improvements**
  - Unified GRN handling, refined financial calculations (per-unit discounts, redistributed proportional allocations, aggregated totals), profit-margin checks, and improved stock/batch updates on finalize.

- **UI**
  - Enhanced GRN receive/costing pages with inline item edits, expenses, preview/print; PO list shows contextual Create/Edit/View GRN actions; admin VTM/VMpp moved to autocomplete and lazy-loading.

### CODE REVIEW

File: src/main/java/com/divudi/bean/pharmacy/GrnCostingController.java [hunk 1]
```diff
@@ -2241,4 +2294,279 @@ private void saveImportBill(Bill importGrn) {
...
+            // Ensure bill discount synchronization before distribution
+            ensureBillDiscountSynchronization();
+            
+            pharmacyCostingService.distributeProportionalBillValuesToItems(getBillItems(), getGrnBill());
+            calculateBillTotalsFromItems();
+            calDifference();
+        }
+        
+        return "/pharmacy/pharmacy_grn_costing_with_save_approve?faces-redirect=true";
+    }
```
[the remark this answers is not in the document]
Comment (buddhika75) [reply src/main/java/com/divudi/bean/pharmacy/GrnCostingController.java:2324]: this will not change distribution. please check again

File: src/main/java/com/divudi/bean/pharmacy/GrnCostingController.java [hunk 2]
```diff
@@ -2241,4 +2294,279 @@
...
```
[the remark this answers is not in the document]
Comment (buddhika75) [reply src/main/java/com/divudi/bean/pharmacy/GrnCostingController.java:2367]: will not affect totals. please recheck

File: src/main/java/com/divudi/bean/pharmacy/GrnCostingController.java [hunk 3]
```diff
@@ -2241,4 +2294,279 @@
...
+        // Final bill update to persist the collection
+        getBillFacade().edit(getCurrentGrnBillPre());
+
+        // Ensure bill discount distribution before saving (even if 0 to clear previous distributions)
+        ensureBillDiscountSynchronization();
+        pharmacyCostingService.distributeProportionalBillValuesToItems(getBillItems(), getGrnBill());
+
+        // Update totals - no copying needed since grnBill and currentGrnBillPre are the same object
+        calculateBillTotalsFromItems();
+        getBillFacade().edit(getCurrentGrnBillPre());
```
[the remark this answers is not in the document]
Comment (buddhika75) [reply src/main/java/com/divudi/bean/pharmacy/GrnCostingController.java:2459]: this is working properly. Please ignore

File: src/main/java/com/divudi/bean/pharmacy/GrnCostingController.java [hunk 4]
```diff
@@ -2241,4 +2294,279 @@ private void saveImportBill(Bill importGrn) {
...
+            getBillFacade().edit(getCurrentGrnBillPre());
+        }
+
+        // Save bill items - work directly with bill's collection
+        for (BillItem i : getCurrentGrnBillPre().getBillItems()) {
+            if (i.getBillItemFinanceDetails() != null && 
+                i.getBillItemFinanceDetails().getQuantity().doubleValue() == 0.0 && 
+                i.getBillItemFinanceDetails().getFreeQuantity().doubleValue() == 0.0) {
+                continue;
+            }
```
[the remark this answers is not in the document]
Comment (buddhika75) [reply src/main/java/com/divudi/bean/pharmacy/GrnCostingController.java]: this is working properly. Please ignore
